bin/propolis-server/src/lib/initializer.rs

```diff
@@ -272,6 +275,104 @@ impl<'a> MachineInitializer<'a> {
         Ok(())
     }
 
+    fn create_storage_backend_from_spec(
+        &self,
+        backend_spec: &instance_spec::v0::StorageBackendV0,
+        backend_name: &str,
+        nexus_client: &Option<NexusClient>,
+    ) -> Result<StorageBackendInstance, Error> {
+        match backend_spec {
+            instance_spec::v0::StorageBackendV0::Crucible(spec) => {
+                info!(self.log, "Creating Crucible disk";
+                      "serialized_vcr" => &spec.request_json);
+
+                let vcr: VolumeConstructionRequest =
+                    serde_json::from_str(&spec.request_json)?;
+
+                let cru_id = match vcr {
+                    VolumeConstructionRequest::Volume { id, .. } => {
+                        id.to_string()
+                    }
+                    VolumeConstructionRequest::File { id, .. } => {
+                        id.to_string()
+                    }
+                    VolumeConstructionRequest::Url { id, .. } => id.to_string(),
+                    VolumeConstructionRequest::Region { .. } => {
+                        "Region".to_string()
+                    }
+                };
+
+                let be = propolis::block::CrucibleBackend::create(
+                    vcr,
+                    spec.readonly,
+                    self.producer_registry.clone(),
+                    nexus_client.clone(),
+                    self.log.new(
+                        slog::o!("component" => format!("crucible-{cru_id}")),
+                    ),
+                )?;
+
+                let child = inventory::ChildRegister::new(
+                    &be,
+                    Some(be.get_uuid()?.to_string()),
+                );
+
+                let crucible = Some((be.get_uuid()?, be.clone()));
+                Ok(StorageBackendInstance { be, child, crucible })
+            }
+            instance_spec::v0::StorageBackendV0::File(spec) => {
+                info!(self.log, "Creating file disk backend";
+                      "path" => &spec.path);
+
+                let nworkers = NonZeroUsize::new(8).unwrap();
+                let be = propolis::block::FileBackend::create(
+                    &spec.path,
+                    spec.readonly,
+                    nworkers,
+                    self.log.new(
+                        slog::o!("component" => format!("file-{}", spec.path)),
+                    ),
+                )?;
+
+                let child =
+                    inventory::ChildRegister::new(&be, Some(spec.path.clone()));
+                Ok(StorageBackendInstance { be, child, crucible: None })
+            }
+            instance_spec::v0::StorageBackendV0::InMemory(spec) => {
+                let bytes = base64::Engine::decode(
+                    &base64::engine::general_purpose::STANDARD,
+                    &spec.base64,
```

Comment:
We discussed this some offline. My general goal here has long been to phase out propolis-server's `instance_ensure` endpoint in favor of the `instance_spec_ensure` endpoint and to use spec elements throughout the stack as a common way of referring to components in a VM. In that world, the initial disk contents would need to be included here, since they're the equivalent of the "cloud-init" base64 string the `instance_ensure` parameters provide today. (At the end of the day it has to get to Propolis somehow or another.)

That said, we do need to be careful about how we display these (and, as you pointed out offline, Crucible VCRs), because the in-memory disks contain user data and Crucible VCRs contain Crucible encryption keys. I'll add the necessary Display/Debug impls to these components before merging.

Reply:
Also discussed was renaming the `InMemory` element to something like `Blob`, since the important aspect of that backend is that Nexus is providing the contents, not that propolis happens to store it in-memory for servicing requests.

Reply:
Both these items (the `Debug` redactions and the renaming of `InMemoryStorageBackend`) should be resolved by c1398c5.

Reply:
That commit clearly wasn't sufficient to the task, but 55d14be should clean up the remaining messes (`InMemory` now only appears in propolis-lib and in the machine initializer line that instantiates a `block::InMemoryBackend`).